### ISSUE TITLE
429 Client Error - API limits in place?

### ISSUE THREAD

**Issue by BinkyWong** [rarcega/instagram-scraper#208]
Awesome script! Really useful 👍 

When scraping, when it hits 100 pictures, it gives this 

`Searching xxxxxx for posts: 100 media [00:32,  3.12 media/s]
Repeated error HTTPError('429 Client Error: - for url: https://www.instagram.com/graphql/query/?query_id=xxxxxxxx&shortcode=xxxxxxx&first=100&after=',)
(A)bort, (I)gnore, (R)etry or retry (F)orever?`

I leave it a while and then hit retry and it goes again for another 100 and stops again. Could we wrap this with a proxy rotator (https://codelike.pro/create-a-crawler-with-rotating-ip-proxy-in-python/) so say, it rotates the proxy after 50 images?

**Comment by VasiliPupkin256** [comment]
If everyone will use the same proxy list, instagram would ban these proxies and everyone will get 429 errors again

**Comment by BinkyWong** [comment]
quite true - however, it can be mitigated by randomizing the proxy list and using custom lists

**Comment by VasiliPupkin256** [comment]
We need some research in the topic to make it act more natural and mimic instagram app better. It bans the scraper process from an ip for an hour. This ruins the batch processing mode.

**Comment by KobaKhit** [comment]
For me I get the same error at 1900 posts.

**Comment by XxbrandoXx** [comment]
Hi, 
The parameter --pause x in milliseconds will be very useful here, (need to be created)
It would be a pause between scraping two users, making a random pause from X to Y milliseconds will be really worth it as it could bypass the security (but I don't think so, need to do some tests).
This could a good improvment but need some work and tries to make sure it is useful.

I am having issues with login but even without being logged in, I'm reaching the same error "HTTPError(u'429 Client Error:XXXX)" and the tool only downloaded 20 pictures (and accessed to 20 private profiles accounts)

**Comment by VasiliPupkin256** [comment]
@XxbrandoXx have you tried this pause strategy? The delay `time.sleep(random.uniform(5, 10))` between users is not enough to overcome the ban

**Comment by XxbrandoXx** [comment]
Hi @VasiliPupkin256 
Thanks for the reply.

I never tried that strategy, but as you're telling me it won't bypass the ban, I won't waste time doing it

The tool maybe need a different way to work.

First look (if possible) which account has a story on, and put the names in a string or array
Then access to these pictures with a delay +- 2 seconds to DL them
Doing that way it'll looks like an human but will only works with the stories, to scrape a user I don't have any idea right now

But I never looked after the Instagram's API so maybe what i'm saying is not possible ^^

**Comment by Moondra** [comment]
I'm running into this error as well. Anyone have any temporary solutions?

**Comment by oferchen** [comment]
by looking at the developer documentation there could be a way to reset the limits counter
renew the session token by logging in again and renew the process whenever there is a 429 code received

**Comment by VasiliPupkin256** [comment]
Relogin doesn't work for me. Changing IP doesn't help either. I am able to partially overcome the issue using many different accounts only.

**Comment by suruisunstat** [comment]
I got this same 429 error after extracting 20 posts:
requests.exceptions.HTTPError: 429 Client Error: - for url: https://www.instagram.com/graphql/query/?query_hash=42323d64886122307be10013ad2dcc44&variables=xxxxfirstxxxxafterxxxx
Has anyone figured out how to deal with this error? I tried with a new login account but it did not work either.

**Comment by VasiliPupkin256** [comment]
@suruisunstat the ban lasts for an hour approximately

**Comment by suruisunstat** [comment]
@VasiliPupkin256 Yes, it worked after an hour. However, I can only succeed in extracting one user at a time. For the second user, the same 429 error pops up again. What I want to do is to extract multiple users within an hour, therefore it is not very helpful, need to figure out how to avoid this error...

**Comment by Greenbald** [comment]
I have the same problem as @XxbrandoXx,  the algorithm only scrapes 20 pictures and then gives the following error : 

2018-06-11 11:42:45,293 - instagram_scraper.app - WARNING - Retry after exception HTTPError('429 Client Error: - for url: https://www.instagram.com/graphql/query/?query_hash=33ba35852cb50da46f5b5e889df7d159&variables=%7B%22shortcode%22:%22BdF1lPWBHX_%22,%22first%22:50,%22after%22:%22%22%7D',) on https://www.instagram.com/graphql/query/?query_hash=33ba35852cb50da46f5b5e889df7d159&variables={"shortcode":"BdF1lPWBHX_","first":50,"after":""}

I'm using the following params :

``` instagram-scraper -f <filename> -d <destination> -n -u <username> -p <password> --retry-forever --latest --coments ```

**Comment by samhains** [comment]
Also looking for temporary solution to this :|

**Comment by samhains** [comment]
its crazy, even with a rolling proxy wrapping every request, and a random 2-4 second wait in between every request - i'm still getting 429's.... how does it know???!?!?!

**Comment by AndCycle** [comment]
`instaloader` do have some measure in-place which is just sleep and wait for a while
which is a argument in `instaloader` called `graphql_rate_limit`

**Comment by VasiliPupkin256** [comment]
@AndCycle instagram-scraper is doing the same basically, just not telling the end user that the cause of errors is rate limit

**Comment by AndCycle** [comment]
@VasiliPupkin256 but not built-in for auto retry, which `instaloader` already done

currently instagram is really sensitive to this kind of API access, 
once I hit this error on insagram-scraper it require around few days until they move me out of the blacklist.

**Comment by VasiliPupkin256** [comment]
@AndCycle this is exactly what scraper is asking in a prompt `Repeated error... (A)bort, (I)gnore, (R)etry or retry (F)orever?` and it also has an option `--retry-forever`, use it

**Comment by VasiliPupkin256** [comment]
@AndCycle I've played with instaloader recently. The heuristic it use is the limit of graphql requests to 20 in 11 minutes. It really slow down the download process but seems to eliminates 429 errors. The batch run that scraper is capable to do in 5 minutes from multiple accounts, is done in like 45 minutes by instaloader. But running from single account scraper took two hours to complete the same batch due to bans and delays. It seems reasonable to implement this limit heuristic though. When the major login bug will be fixed, I'll implement this heuristic here

**Comment by thewracl** [comment]
@VasiliPupkin256 
To be honest, after it seems @rarcega lost complete interest in instagram-scraper for whatever reasons, it would be better to try to make either a fork of instaloader or help to improve it. At least this works. But I have a few minor things I don't like: For one, it has no option for using the filename Instagram is using, f.e. 18258718_215174558973293_6313767881216372736_a.jpg and I also don't like when it downloads all visible stories every now and then. I have a strict folder system so these stories interfere with my system because I have the wrong files in the directory. I just can use :stories when I need it ...

**Comment by VasiliPupkin256** [comment]
It seams than instagram have raised the graphgl requests limit from 20 to 200. @AndCycle @thewracl  can you confirm that the issue now appear less frequently? I need to know what constant should I use.

**Comment by AndCycle** [comment]
Instagram just make the limit back to 20, so there need to have a fuzzy code to do the retry, or the script will just bail out.

**Comment by igorveremsky** [comment]
Any progress on fixing this issue? Can it be connected with some adjustments requests that real browser can done? For real, users in less than 5 minutes can scroll more than 20 pages of timelines.

**Comment by igorveremsky** [comment]
@VasiliPupkin256 it exist in my own scraper that have same scrap logic as this extension

**Comment by paullovachy** [comment]
Having the same issue!

**Comment by santosadrian** [comment]
Repeated error HTTPError(u'429 Client Error: - for url: bla,')
Still getting the same issue, but I downloaded arround 20 profiles, so the solution is login with a different user when the error happens?
